### PR TITLE
Maassi (Example Bot in Python)

### PR DESCRIPTION
**Pull request by jankratochvilcz**
👋

I'm trying to learn Python a bit and I thought it could be useful if I fill out our samples gallery a bit in the process. Maassi (Mildly Annoyed ASSIstant) is a bot built with Python and RQ (for async processing). My goal with the bot is to add the "features I with Twist had" for personal use as well as to explore the limitations of ourAPI.

Would appreciate a review, I'm sure I made a ton of rookie mistakes during the implementation 😅

Some notes from my experience (no need to comment, I will make a post where sum things up when I have enough experience with the API):
* Twist returns `500` when a JSON payload is incorrectly formatted. A more informative response would make the troubleshooting easier
* The 10s timeout is mentioned [under slash integrations](https://developer.twist.com/v3/#slash-commands) but seems to apply to bots as well. Its is confusing
* It's not clear that a bot integration is a "general integration"
* Generally the I would love a doc page on "here's how you build a bot", it's a bit strange that it's somewhat randomly placed in the reference documentation

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
-node_modules
-.vscode
-.DS_Store
\ No newline at end of file
+node_modules
+.vscode
+.DS_Store
+**/__pycache__/**
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply .gitignore]: Moved to my global .gitignore, thanks!

File: bot_integration/python/app.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+from rq import Queue
+
+from model import CommandType, CommandPayload
+
+from commands.stale_threads import getStaleThreads
+from commands.unfollow import unfollow
+from commands.help import help
+
+from twist_model import workspaceIdParamName, userIdParamName, urlCallbackParamName, contentParameterName
+from output_formatting import startingWorkerMessage, unknownCommandMessage, cannotUnfollowMessage, unfollowedMessage
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/app.py]: Thanks for the tip, I've done this. I didn't use all of them, specifically the one for reordering imports seemed to be incompatible with Black. I tried to correct the files quite a few times but during pre-commit, they kept getting "corrected" with changes that were themselves violations, maybe I'm missing something :-/

File: bot_integration/python/readme.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Stale Issues - Bot Integration with Twist
+
+This sample shows a simple integration with the [Bot](https://developer.twist.com/v3/#how-a-bot-works) endpoint of Twist. It receives a message, gets data from the Twist API and returns them back to the user. It demonstrates the following:
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/readme.md]: Changed to a more generic link :-)

File: bot_integration/python/readme.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+2. Run `flask run` to start the Flask web server
+3. Run `rq worker` to start the worker that the Flask web server hands of the requests to. (if you see a "Connection refused" message, Redis is most likely not running. See below for resolution)
+
+To install [Ngrok](https://ngrok.com/), you need to register and then install the binaries through the [setup & installation](https://dashboard.ngrok.com/get-started) steps.
+
+Ensure that Redis server is installed and it's running using `redis-cli ping` (you should see a `PONG` response). If you don't get a response, try running `sudo service redis-server start` on bash and ping again.
+
+In addition, you need to register your integration with Twist. You can do this [here](https://twist.com/integrations/build). Pick the General Integration type and write the name *Stale Threads*. After the integration is created, go to the *OAuth Authentication* section, copy the *OAuth2 test token* and paste it in `twist_api.py`. Then go to the *Bot* tab and set the *Outgoing webhook URL* to `https://[Ngrok HTTPS Forward URL]/stale-threads`, where the Ngrok URL is the one you saw when running `ngrok http 5000` earlier.
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/readme.md]: Went with the last solution, thanks for the suggestion!

[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/readme.md]: Have made this clearer :-)

File: bot_integration/python/twist_api.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+from requests import get, post
+from datetime import timedelta, datetime
+from twist_model import workspaceIdParamName, archivedParamName, idParamName, channelIdParamName, closedParamName, creatorParamName, lastUpdatedParamName
+
+twistToken = 'oauth2:a3d36df4460d49c46c11afadeb7b4600321200de'
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/twist_api.py]: Sorry about this, it snuck in when doing changes :-/ (the token is already invalidated and have removed this)

File: bot_integration/python/readme.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+Finally, you need to go to the *Installation 🚀* tab and install the integration either into your workspace or sharing it with someone else.
+
+Now, you should be ready to go to the workspace where you installed the integration, find a user that's named *Stale Threads*. Write this user a new message with the content `stale threads` and wait a bit. You should first see the message of *Getting stale threads. This will take a couple of seconds.* and ten to twenty seconds later, the list of your stale threads should pop up.
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/readme.md]: Have added clarity

File: bot_integration/python/readme.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+### Windows
+
+Please note that for Windows, you need to run the sample through the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10). The reason is that RQ has a dependancy on Redis, which doesn't natively run on Windows. To setup Ngrok on WSL, grab the linux link from [here](https://dashboard.ngrok.com/get-started) and run the following from WSL (Ubuntu):
+* `sudo apt get unzip` - installs an unzip tool
+* `wget [link URL]` - downloads the zip with binaries
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply bot_integration/python/readme.md]: It's in the previous sentence, do you find it insufficiently clear?

File: .gitignore [hunk 8]
```diff
@@ -1,3 +1,5 @@
-node_modules
-.vscode
-.DS_Store
\ No newline at end of file
+node_modules
+**/__pycache__/**
```
Comment (PotHix) [inline .gitignore]: Are the double asterisks here really needed?

File: .gitignore [hunk 9]
```diff
@@ -1,3 +1,5 @@
-node_modules
-.vscode
-.DS_Store
\ No newline at end of file
+node_modules
+**/__pycache__/**
+lib
+bin
```
Comment (PotHix) [inline .gitignore]: It's weird to have `lib` and `bin` here. Any reason?

File: README.md [hunk 10]
```diff
@@ -27,4 +27,4 @@ Examples
 * [Appear.in slash integration](slash_integration/node/)
 
 ### Go
-* [Appear.in slash integration](slash_integration/go/)
\ No newline at end of file
+* [Appear.in slash integration](slash_integration/go/)
```
Comment (PotHix) [inline README.md]: We should probably change that to Whereby, since Appear.in changed its name.

File: bot_integration/python/app.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    workspaceIdParamName,
+)
+
+app = Flask(__name__)
+
+# We get a reference to the queue on redis and add a new task to it.
+queue = Queue(connection=Redis())
+
+
+def processStaleThreads(commandPayload):
```
Comment (PotHix) [inline bot_integration/python/app.py]: Please refer to the naming conventions: https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

Function names should be lowercase, with words separated by underscores as necessary to improve readability.

Please review all instances of methods and variables.

File: bot_integration/python/app.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def processUnrecognized(commandPayload):
+    return jsonify({contentParameterName: unknownCommandMessage})
+
+
+def processUnfollow(commandPayload):
+    canUnfollow = unfollow(commandPayload.commandParameter)
+
+    if canUnfollow:
```
Comment (PotHix) [inline bot_integration/python/app.py]: I suggest to use if to set a message, so you can `jsonify` only once. Example:

```
if canUnfollow:
    message = unfollowedMessage
else:
    message = cannotUnfollowMessage

return jsonify({contentParameterName: message})
```

File: bot_integration/python/readme.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (PotHix) [reply bot_integration/python/readme.md]: Yes, I do. I prefer to repeat things to make it clear when dealing with documentation (or tests)

File: bot_integration/python/readme.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+Finally, you need to go to the *Installation 🚀* tab and install the integration either into your workspace or sharing it with someone else.
+
+Now, you should be ready to go to the workspace where you installed the integration, find a user that's named *Stale Threads*. Write this user a new message with the content `stale threads` and wait a bit. You should first see the message of *Getting stale threads. This will take a couple of seconds.* and ten to twenty seconds later, the list of your stale threads should pop up.
+
+### Windows
+
+Please note that for Windows, you need to run the sample through the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10). The reason is that RQ has a dependancy on Redis, which doesn't natively run on Windows. To setup Ngrok on WSL, grab the linux link from [here](https://dashboard.ngrok.com/get-started) and run the following from WSL (Ubuntu):
+* `sudo apt get unzip` - installs an unzip tool
```
[the remark this answers is not in the document]
Comment (PotHix) [reply bot_integration/python/readme.md]: Is this really resolved?

File: bot_integration/python/command_processing.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+
+def processInput(input):
+    matchingCommands = [
+        member
+        for name, member in CommandType.__members__.items()
+        if input.startswith(member.value)
+    ]
+
+    if len(matchingCommands) == 0:
```
Comment (lefcha) [inline bot_integration/python/command_processing.py]: Can be also done (and usually is done in Python) like this:

```python
if matchingCommands:
```

File: bot_integration/python/output_formatting.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    + "* `help` - See what I can do for you."
+)
+
+
+# Outputs a string similar to
+# "[Weekly Snippets](twist URL) last updated on Thursday, February 06"
+def formatStaleThreadString(thread):
+    utcDate = datetime.datetime.utcfromtimestamp(thread[lastUpdatedParamName])
+
+    return (
```
Comment (lefcha) [inline bot_integration/python/output_formatting.py]: It's a bit hard to follow what this outputs, even though you give an example at the function comment.

I would probably break this somehow...

```python
return "[%s](%s) last updated on %s" % (
    thread[titleParamName]}],
    getThreadUrl(
    	thread[workspaceIdParamName], 
        thread[channelIdParamName],
        thread[idParamName]
    ),
    utcDate.strftime('%A, %B %d')
)
```

**Comment by lefcha** [comment]
The file `bot_integration/python/requirements.txt` shows up as binary and doesn't show itself in the review:
Screenshot from 2020-03-09 18-36-09

Maybe it has some text character at the end? https://github.com/Doist/twist-integration-examples/blob/92d7a9382d87bba12cca1b646cec0707678f7653/bot_integration/python/requirements.txt

Also, are the newline chars correct, because it shows a lot of blank lines.

**Comment by imankulov** [comment]
I didn't have a close look, but somehow Maassi polluted Sentry with a bunch of error messages (https://sentry.todoist.net/doist/twist-server/issues/296859/?query=is%3Aunresolved).

At first sight, the Twist API accepted the message content being a list of strings (instead of a single string) and passed it through until the job got stuck into the queue. Since you're onto this, can you please identify the core issue, and maybe create a PR to fix it.

I can also agree with what Jan would say "if we'd use here a [language](https://dotnet.microsoft.com/download/dotnet-core) with proper static typing, this would never happen in the first place."

**Comment by jankratochvilcz** [comment]
I'm abandoning this PR as I don't have the time to make it great, and it would probably also not be a great example as it's very complex at this point 😅
